### PR TITLE
ENH: Add Styler.where

### LINKED ISSUE #16255: API: Add Styler.where
#### Code Sample, a copy-pastable example if possible

```python
# Instead of 
df.style.applymap(lambda val: 'color: %s' % 'red' if val < 0 else 'black')
```

It would be

```python
df.style.where(lambda val: val < 0, 'color: red', 'color: black')
```

### PR DESCRIPTION
**Pull request by Licht-T**
- [x] closes #16255
- [x] tests added / passed
Added `test_where_subset` in `pandas/tests/io/formats/test_style.py`.
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry
It makes the redundant style condition simple.
```python
df.style.applymap(lambda val: 'color: %s' % 'red' if val < 0 else 'black')
df.style.where(lambda val: val < 0, 'color: red', 'color: black')
```

### CODE REVIEW

File: pandas/io/formats/style.py [hunk 1]
```diff
@@ -623,6 +623,32 @@ def applymap(self, func, subset=None, **kwargs):
                            (func, subset), kwargs))
         return self
 
+    def where(self, func, style1, style2='', subset=None, **kwargs):
+        """
+        Apply a function elementwise, updating the HTML
+        representation with a style which is selected in
+        accordance with the return value of a function.
```
Comment (jreback) [inline pandas/io/formats/style.py]: add a versionadded tag

File: pandas/io/formats/style.py [hunk 2]
```diff
@@ -623,6 +623,32 @@ def applymap(self, func, subset=None, **kwargs):
                            (func, subset), kwargs))
         return self
 
+    def where(self, func, style1, style2='', subset=None, **kwargs):
+        """
```
Comment (jreback) [inline pandas/io/formats/style.py]: defaults should be ``None`` not ``''``

File: pandas/io/formats/style.py [hunk 3]
```diff
@@ -623,6 +623,32 @@ def applymap(self, func, subset=None, **kwargs):
...
+        Apply a function elementwise, updating the HTML
+        representation with a style which is selected in
+        accordance with the return value of a function.
+
+        Parameters
+        ----------
+        func : function
```
Comment (jreback) [inline pandas/io/formats/style.py]: callable

File: pandas/io/formats/style.py [hunk 4]
```diff
@@ -623,6 +623,32 @@ def applymap(self, func, subset=None, **kwargs):
...
+            ``func`` should take a scalar and return a boolean
+        style1 : a style applied when ``func`` returns true
```
Comment (jreback) [inline pandas/io/formats/style.py]: style1: str
    when applied......

File: pandas/tests/io/formats/test_style.py [hunk 5]
```diff
@@ -265,6 +265,26 @@ def f(x):
                             col in self.df.loc[slice_].columns)
             assert result == expected
 
+    def test_where_subset(self):
+        def f(x):
```
Comment (jreback) [inline pandas/tests/io/formats/test_style.py]: add a comment with the issue number (this PR number)

**Review by jreback — CHANGES_REQUESTED** [review]
can you add a whatsnew note on Other Enahancements as well as use this in the notebook where appropriate

File: pandas/tests/io/formats/test_style.py [hunk 6]
```diff
@@ -265,6 +265,26 @@ def f(x):
...
+            return style1 if f(x) else style2
+
+        slices = [pd.IndexSlice[:], pd.IndexSlice[:, ['A']],
+                  pd.IndexSlice[[1], :], pd.IndexSlice[[1], ['A']],
+                  pd.IndexSlice[:2, ['A', 'B']]]
+
+        for slice_ in slices:
+            result = self.df.style.where(f, style1, style2,
+                                         subset=slice_)._compute().ctx
+            expected = self.df.style.applymap(g, subset=slice_)._compute().ctx
```
Comment (TomAugspurger) [inline pandas/tests/io/formats/test_style.py]: These tests are fine, but could you add one test that checks the generated values directly against the dictionary?

**Comment by Licht-T** [comment]
@jreback @TomAugspurger Thanks for your review. Fixed and made some new tests.

**Comment by Licht-T** [comment]
Travis CI seems not work well due to its own error.
Could someone restart the Travis CI build?

File: doc/source/whatsnew/v0.21.0.txt [hunk 7]
```diff
@@ -112,6 +112,7 @@ Other Enhancements
 - `read_*` methods can now infer compression from non-string paths, such as ``pathlib.Path`` objects (:issue:`17206`).
 - :func:`pd.read_sas()` now recognizes much more of the most frequently used date (datetime) formats in SAS7BDAT files (:issue:`15871`).
 - :func:`DataFrame.items` and :func:`Series.items` is now present in both Python 2 and 3 and is lazy in all cases (:issue:`13918`, :issue:`17213`)
+- :func:`Styler.where` has been implemented. It enables simple DataFrame styling on the Jupyter notebook (:issue:`17474`).
```
Comment (jreback) [inline doc/source/whatsnew/v0.21.0.txt]: pls label this as a convenience for Styler.applymap

File: pandas/io/formats/style.py [hunk 8]
```diff
@@ -623,6 +623,40 @@ def applymap(self, func, subset=None, **kwargs):
...
+        """
+        Apply a function elementwise, updating the HTML
+        representation with a style which is selected in
+        accordance with the return value of a function.
+
+        .. versionadded:: 0.21.0
+
+        Parameters
+        ----------
+        func : callable
```
Comment (jreback) [inline pandas/io/formats/style.py]: this should mimic the signature of ``DataFrame.where`` as much as possible

func -> cond
style1 -> value
style2 -> other

**Review by jreback — APPROVED** [review]
this seems ok, @TomAugspurger . do we have a section in api.rst for things like this?

**Comment by Licht-T** [comment]
@jreback Thanks for your suggestion. Fixed!

File: doc/source/whatsnew/v0.21.0.txt [hunk 9]
```diff
@@ -112,6 +112,7 @@ Other Enhancements
 - `read_*` methods can now infer compression from non-string paths, such as ``pathlib.Path`` objects (:issue:`17206`).
 - :func:`pd.read_sas()` now recognizes much more of the most frequently used date (datetime) formats in SAS7BDAT files (:issue:`15871`).
 - :func:`DataFrame.items` and :func:`Series.items` is now present in both Python 2 and 3 and is lazy in all cases (:issue:`13918`, :issue:`17213`)
+- :func:`Styler.where` has been implemented. It is as a convenience for Styler.applymap and enables simple DataFrame styling on the Jupyter notebook (:issue:`17474`).
```
Comment (jreback) [inline doc/source/whatsnew/v0.21.0.txt]: use ``:func:`Styler.applymap` `` as well

File: pandas/io/formats/style.py [hunk 10]
```diff
@@ -623,6 +623,40 @@ def applymap(self, func, subset=None, **kwargs):
...
+            applied when ``cond`` returns false
+        subset : IndexSlice
+            a valid indexer to limit ``data`` to *before* applying the
+            function. Consider using a pandas.IndexSlice
+        kwargs : dict
+            pass along to ``cond``
+
+        Returns
+        -------
+        self : Styler
```
Comment (jreback) [inline pandas/io/formats/style.py]: can you add a ``See Also`` here (pointing to .applymap) and the reciprical one there

File: doc/source/whatsnew/v0.21.0.txt [hunk 11]
```diff
@@ -112,6 +112,7 @@ Other Enhancements
...
```
Comment (jreback) [inline doc/source/whatsnew/v0.21.0.txt]: ``.where`` needs to be added to ``api.rst``

**Review by jreback — CHANGES_REQUESTED** [review]

**Comment by Licht-T** [comment]
@jreback Thanks. All is now fixed.

**Review by TomAugspurger — APPROVED** [review]

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks @Licht-T 

If you'd have a pass thru the actual documentation notebook and make a PR for anything where we should logically use ``.where`` to make the syntax more clear would be great (or indicate that it looks good).

**Pull request merged**